extensions/amp-social-share/1.0/social-share.js

```diff
@@ -0,0 +1,210 @@
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as CSS from './social-share.css';
+import * as Preact from '../../../src/preact';
+import {Keys} from '../../../src/utils/key-codes';
+import {SocialShareIcon} from '../../../third_party/optimized-svg-icons/social-share-svgs';
+import {addParamsToUrl, parseQueryString} from '../../../src/url';
+import {dict} from '../../../src/utils/object';
+import {getSocialConfig} from './amp-social-share-config';
+import {isObject} from '../../../src/types';
+import {openWindowDialog} from '../../../src/dom';
+import {useResourcesNotify} from '../../../src/preact/utils';
+
+const DEFAULT_WIDTH = 60;
+const DEFAULT_HEIGHT = 44;
+
+/**
+ * @param {!JsonObject} props
+ * @return {PreactDef.Renderable}
+ */
+export function SocialShare(props) {
+  const name = 'SocialShare';
+  useResourcesNotify();
+
+  const {
+    'typeConfig': typeConfig,
+    'baseEndpoint': baseEndpoint,
+    'checkedWidth': checkedWidth,
+    'checkedHeight': checkedHeight,
+  } = checkProps(props, name);
+  const finalEndpoint = createEndpoint(typeConfig, baseEndpoint, props);
+
+  /**
+   * @private
+   * @param {!JsonObject} typeConfig
+   * @param {?string} baseEndpoint
+   * @param {!JsonObject} props
+   * @return {?string}
+   */
+  function createEndpoint(typeConfig, baseEndpoint, props) {
+    const {params, bindings} = props;
+    const combinedParams = dict();
+    Object.assign(combinedParams, typeConfig['defaultParams'], params);
+    const endpointWithParams = addParamsToUrl(baseEndpoint, combinedParams);
+
+    const combinedBindings = dict();
+    const bindingVars = typeConfig['bindings'];
+    if (bindingVars) {
+      bindingVars.forEach((name) => {
+        combinedBindings[name.toUpperCase()] = params[name] || '';
+      });
+    }
+    if (bindings) {
+      Object.keys(bindings).forEach((name) => {
+        combinedBindings[name.toUpperCase()] = bindings[name] || '';
+      });
+    }
+    const finalEndpoint = Object.keys(combinedBindings).reduce(
+      (endpoint, binding) =>
+        endpoint.replace(new RegExp(binding, 'g'), combinedBindings[binding]),
+      endpointWithParams
+    );
+    return finalEndpoint;
+  }
+
+  /**
+   * @private
+   */
+  function handleActivation() {
+    const protocol = finalEndpoint.split(':', 1)[0];
+    const windowFeatures = 'resizable,scrollbars,width=640,height=480';
+    if (protocol === 'navigator-share') {
+      if (window && window.navigator && window.navigator.share) {
+        const dataStr = finalEndpoint.substr(finalEndpoint.indexOf('?'));
```

Comment:
To reiterate, what if there's no `?`? What if there's `#`? `praseUrl().queryString` in the `url.js` takes care of all of those issues.

Reply:
Thanks for the calling this out.  I was only able to find parseUrlDeprecated which I am hesitant to use because of the Deprecation.  I have attempted to write my own helper function (getQueryString, line 193), which handles the corner cases that I can think of.

1. No `?` -> returns '' (empty string)
2. `?` and `#` -> only returns the substring from `?` to `#`
3. `#` then `?` (reversed order) -> returns substring from `?` to the end of string
4. `?` and no `#`-> returns substring from `?` to the end of string

Reply:
Let's dedup to https://github.com/ampproject/amphtml/pull/28137#discussion_r424734753.